File: NurSamplePC_vs2008_VB/NurSamplePC_VB/NurSerializer.vb
Imports System.Collections.Generic
Imports System.Text
Imports System.IO
Imports System.Xml.Serialization
Imports NurApiDotNet

Public Class NurSerializer
	Public Structure NurInfo
		Public NurApi As String
		Public NurApiDotNet As String
		Public mode As Char
		Public readerInfo As NurApi.ReaderInfo
		Public moduleVersions As NurApi.ModuleVersions
		Public FWINFO As String
		Public deviceCapabilites As NurApi.DeviceCapabilites
		Public moduleSetup As NurApi.ModuleSetup
		Public sensorConfig As NurApi.SensorConfig
		Public gpioEntry As NurApi.GpioEntry()
		Public gpioStatus As NurApi.GpioStatus()
		Public ethConfig As NurApi.EthConfig
		Public regionInfo As NurApi.RegionInfo()
		Public customHoptable As NurApi.CustomHoptable
		Public customHoptableEx As NurApi.CustomHoptableEx
		Public AntennaInfo As String()
	End Structure

    Private serNurInfo As NurInfo
    Private serializer As XmlSerializer

    ''' <summary>
    ''' Serializes this instance.
    ''' </summary>
    Public Sub Serialize(ByVal hNur As NurApi)
        serNurInfo = New NurInfo()
        Try
            serNurInfo.NurApi = hNur.GetFileVersion()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.NurApiDotNet = NurUtils.NurApiDotNetVersion
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.mode = hNur.GetMode()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.readerInfo = hNur.GetReaderInfo()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.moduleVersions = hNur.GetVersions()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.FWINFO = hNur.GetFWINFO()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.deviceCapabilites = hNur.GetDeviceCaps()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.moduleSetup = hNur.GetModuleSetup()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.sensorConfig = hNur.GetSensorConfig()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.gpioEntry = hNur.GetGPIOConfig()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.gpioStatus = New NurApi.GpioStatus(serNurInfo.readerInfo.numGpio - 1) {}
            For i As Integer = 0 To serNurInfo.gpioStatus.Length - 1
                serNurInfo.gpioStatus(i) = hNur.GetGPIOStatus(i)
            Next
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.ethConfig = hNur.GetEthConfig()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.regionInfo = New NurApi.RegionInfo(serNurInfo.readerInfo.numRegions - 1) {}
            For i As Integer = 0 To serNurInfo.readerInfo.numRegions - 1
                serNurInfo.regionInfo(i) = hNur.GetRegionInfo(i)
            Next
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.customHoptable = hNur.GetCustomHoptable()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.customHoptableEx = hNur.GetCustomHoptableEx()
        Catch generatedExceptionName As NurApiException
        End Try
        Try
            serNurInfo.AntennaInfo = MeasureReflectedPowers(hNur)
        Catch generatedExceptionName As NurApiException
        End Try

        serializer = New XmlSerializer(GetType(NurInfo))

    End Sub

    ''' <summary>
    ''' Measures the reflected powers.
    ''' </summary>
    ''' <param name="hNur">The NurApi.</param>
    ''' <returns></returns>
    Private Function MeasureReflectedPowers(ByVal hNur As NurApi) As String()
        Dim antennaList As New List(Of String)()
        Try
            ' Measure Reflected Power
            Dim antenna As Integer = 0
            Dim mask As Integer = hNur.AntennaMask
            If mask > 0 Then
                Dim tmpSelectdeAntenna As Integer = hNur.SelectedAntenna
                While mask > 0
                    If (mask And 1) <> 0 Then
                        Try
                            ' Select Antenna
                            hNur.SelectedAntenna = antenna
                            ' Measure Reflected Power
                            Dim regionInfo As NurApi.RegionInfo = hNur.GetRegionInfo()
                            hNur.GetReflectedPower(regionInfo.baseFrequency)
                            ' FW bug fix
                            Dim rpInfo As NurApi.ReflectedPowerInfo = hNur.GetReflectedPower()
                            ' Calculate Reflected Power
                            Dim rf As Double = Math.Sqrt(CDbl(rpInfo.iPart * rpInfo.iPart + rpInfo.qPart * rpInfo.qPart))
                            rf /= CDbl(rpInfo.div)
                            rf = Math.Log10(rf) * 20.0
                            If [Double].IsInfinity(rf) Then
                                rf = -30
                            End If
                            If rf < 0 Then
                                antennaList.Add(String.Format("Antenna {0} = Reflected Power {1:0.0} dBm", antenna, rf))
                            Else
                                antennaList.Add(String.Format("Antenna {0} = Reflected Power {1:0.0} dBm. Make sure that the antenna is connected.", antenna, rf))
                            End If
                        Catch generatedExceptionName As NurApiException
                            antennaList.Add(String.Format("Antenna {0} = Could not measure the Reflected Power.", antenna))
                        End Try
                    End If
                    mask = mask >> 1
                    antenna += 1
                End While
                ' Restore SelectedAntenna
                hNur.SelectedAntenna = tmpSelectdeAntenna
            Else
                antennaList.Add("No eabled antennas!")
            End If
        Catch generatedExceptionName As NurApiException
        End Try
        Return antennaList.ToArray()
    End Function

    ''' <summary>
    ''' Saves to file.
    ''' </summary>
    ''' <param name="path">The path.</param>
    Public Sub SaveToFile(ByVal path As String)
        Dim textWriter As TextWriter = New StreamWriter(path)
        serializer.Serialize(textWriter, serNurInfo)
        textWriter.Close()
    End Sub

    ''' <summary>
    ''' Saves to stream.
    ''' </summary>
    ''' <param name="stream">The stream.</param>
    Public Sub SaveToStream(ByVal stream As Stream)
        serializer.Serialize(stream, serNurInfo)
    End Sub
End Class

File: WindowsCE/NurSample_vs2008/NurSample/NurSerializer.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Xml.Serialization;
using NurApiDotNet;

namespace NurSample
{
    public class NurSerializer
    {
        public struct NurInfo
        {
            public string NurApi;
            public string NurApiDotNet;
            public char mode;
            public NurApi.ReaderInfo readerInfo;
            public NurApi.ModuleVersions moduleVersions;
            public string FWINFO;
            public NurApi.DeviceCapabilites deviceCapabilites;
            public NurApi.ModuleSetup moduleSetup;
            public NurApi.SensorConfig sensorConfig;
            public NurApi.GpioEntry[] gpioEntry;
            public NurApi.GpioStatus[] gpioStatus;
            public NurApi.EthConfig ethConfig;
            public NurApi.RegionInfo[] regionInfo;
            public NurApi.CustomHoptable customHoptable;
            public NurApi.CustomHoptableEx customHoptableEx;
            public string[] AntennaInfo;
        }

        NurInfo nurInfo;
        XmlSerializer serializer;

        /// <summary>
        /// Serializes this instance.
        /// </summary>
        public void Serialize(NurApi hNur)
        {
            nurInfo = new NurInfo();
            try
            {
                nurInfo.NurApi = hNur.GetFileVersion();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.NurApiDotNet = NurUtils.NurApiDotNetVersion;
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.mode = hNur.GetMode();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.readerInfo = hNur.GetReaderInfo();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.moduleVersions = hNur.GetVersions();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.FWINFO = hNur.GetFWINFO();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.deviceCapabilites = hNur.GetDeviceCaps();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.moduleSetup = hNur.GetModuleSetup();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.sensorConfig = hNur.GetSensorConfig();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.gpioEntry = hNur.GetGPIOConfig();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.gpioStatus = new NurApi.GpioStatus[nurInfo.readerInfo.numGpio];
                for (int i = 0; i < nurInfo.gpioStatus.Length; i++)
                {
                    nurInfo.gpioStatus[i] = hNur.GetGPIOStatus(i);
                }
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.ethConfig = hNur.GetEthConfig();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.regionInfo = new NurApi.RegionInfo[nurInfo.readerInfo.numRegions];
                for (int i = 0; i < nurInfo.readerInfo.numRegions; i++)
                {
                    nurInfo.regionInfo[i] = hNur.GetRegionInfo(i);
                }
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.customHoptable = hNur.GetCustomHoptable();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.customHoptableEx = hNur.GetCustomHoptableEx();
            }
            catch (NurApiException) { }
            try
            {
                nurInfo.AntennaInfo = MeasureReflectedPowers(hNur);
            }
            catch (NurApiException) { }

            serializer = new XmlSerializer(typeof(NurInfo));

        }

        /// <summary>
        /// Measures the reflected powers.
        /// </summary>
        /// <param name="hNur">The NurApi.</param>
        /// <returns></returns>
        string[] MeasureReflectedPowers(NurApi hNur)
        {
            List<string> antennaList = new List<string>();
            try
            {
                int tmpSelectdeAntenna = hNur.SelectedAntenna;
                uint tmpAntennaMaskEx = hNur.AntennaMaskEx;

                bool tuneEventsEnabled = hNur.EnableTuneEvents;
                if (tuneEventsEnabled)
                    hNur.EnableTuneEvents = false;

                // Measure Reflected Power
                foreach (string physicalAntenna in hNur.AvailablePhysicalAntennas)
                {
                    try
                    {
                        // Select Antenna
                        hNur.EnablePhysicalAntenna(physicalAntenna);
                        int antennaId = hNur.NurPhysicalAntennaToAntennaId(physicalAntenna);
                        hNur.SelectedAntenna = antennaId;
                        // Measure Reflected Power
                        double dBm = hNur.GetReflectedPowerValue(0);
                        if (dBm < 0)
                        {
                            antennaList.Add(string.Format("{0}:{1} = Reflected Power {2:0.0} dBm", antennaId, physicalAntenna, dBm));
                        }
                        else
                        {
                            antennaList.Add(string.Format("{0}:{1} = Reflected Power {2:0.0} dBm. This antenna may need service.", antennaId, physicalAntenna, dBm));
                        }
                    }
                    catch (NurApiException)
                    {
                        antennaList.Add(string.Format("{0} = Could not measure the Reflected Power.", physicalAntenna));
                    }
                }

                // Restore SelectedAntenna
                hNur.AntennaMaskEx = tmpAntennaMaskEx;
                hNur.SelectedAntenna = tmpSelectdeAntenna;
                if (tuneEventsEnabled)
                    hNur.EnableTuneEvents = true;
            }
            catch (NurApiException)
            {
            }
            return antennaList.ToArray();
        }

        /// <summary>
        /// Saves to file.
        /// </summary>
        /// <param name="path">The path.</param>
        public void SaveToFile(string path)
        {
            TextWriter textWriter = new StreamWriter(path);
            serializer.Serialize(textWriter, nurInfo);
            textWriter.Close();
        }

        /// <summary>
        /// Saves to stream.
        /// </summary>
        /// <param name="stream">The stream.</param>
        public void SaveToStream(Stream stream)
        {
            serializer.Serialize(stream, nurInfo);
        }
    }
}
